aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs

```diff
@@ -0,0 +1,196 @@
+            let output_dir = output_path.unwrap_or_else(|| {
+                format!(
+                    "expected_db_output_files/imported_testnet_txns/{}/{}_{}.json",
+                    processor_name, processor_name, txn_version
+                )
+            });
+
+            // Extract the directory path from the full file path
+            if let Some(parent_dir) = Path::new(&output_dir).parent() {
+                // Create the directory if it doesn't exist
+                fs::create_dir_all(parent_dir)?;
+            }
+
+            // Save the database output as JSON to the specified path
+            fs::write(output_dir.clone(), serde_json::to_string_pretty(&db_values)?)?;
+            println!("[INFO] Generated output file at: {}", output_dir);
+        } else {
+            println!("[INFO] Skipping file generation as requested.");
+        }
+
+        Ok(db_values)
+    }
+
+    /// Helper function to set up and run the mock GRPC server.
+    async fn setup_mock_grpc(&self, transactions: Vec<TransactionsResponse>, chain_id: u64) {
+        println!("received transactions size: {:?}", transactions.len());
+        let mock_grpc_server = MockGrpcServer {
+            transactions,
+            chain_id,
+        };
+
+        // Start the Mock GRPC server
+        tokio::spawn(async move {
+            println!("Starting Mock GRPC server");
+            mock_grpc_server.run().await;
+        });
+    }
+
+    pub fn create_transaction_stream_config(&self,
+        starting_version: u64,
+        ending_version: u64,
+    ) -> TransactionStreamConfig {
+        TransactionStreamConfig {
+            indexer_grpc_data_service_address: Url::parse("http://localhost:51254")
+                .expect("Could not parse database url"),
+            starting_version: Some(starting_version), // dynamically pass the starting version
+            request_ending_version: Some(ending_version), // dynamically pass the ending version
+            auth_token: "".to_string(),
+            request_name_header: "sdk testing".to_string(),
+            indexer_grpc_http2_ping_interval_secs: 30,
+            indexer_grpc_http2_ping_timeout_secs: 10,
+            indexer_grpc_reconnection_timeout_secs: 10,
+            indexer_grpc_response_item_timeout_secs: 60,
+        }
+    }
+
+}
+
+#[async_trait]
+pub trait TestDatabase: Send + Sync {
+    /// Set up the test container using user-defined code.
+    async fn setup<'a>(&'a mut self) -> anyhow::Result<()>;
+
+    /// Retrieve the database connection URL after setup.
+    fn get_db_url(&self) -> String;
+}
+
+pub struct PostgresTestDatabase {
+    connection_string: String,
+    postgres_container: Option<ContainerAsync<GenericImage>>,
+}
+
+impl PostgresTestDatabase {
+    pub fn new() -> Self {
+        PostgresTestDatabase {
+            postgres_container: None,
+            connection_string: String::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl TestDatabase for PostgresTestDatabase {
```

Comment:
Since cargo runs tests in parallel by default, do we need to worry about dirty state in the testcontainers? Maybe we can pass in an additional optional `db_name` param to setup to have different DB names to ensure isolation?

Reply:
Good point!!!

I think,  in current setup using `testcontainers`, each test case will run a separate database instance, because the `PostgresTestDatabase::setup` method starts a new Postgres container for each test. This should ensure that the tests are isolated at the container level, and each test will have its own instance of the database. but let me double-check i may be wrong lol

Reply:
Sounds good. I have trauma from this in the past lol

Reply:
haha me too lol. i added another test case, and fortunately not an issue!

Reply:
awesome thanks!